File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool isValidBST(TreeNode* root, long lowLimit, long highLimit) {
        if(!root) return true;
        if(root->val <= lowLimit || root->val >= highLimit) return false;
        return isValidBST(root->left, lowLimit, root->val) && isValidBST(root->right, root->val, highLimit);
    }
public:
    bool isValidBST(TreeNode* root) {
        return isValidBST(root, LONG_MIN, LONG_MAX);
    }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def find(self, root, left, right): 
        if root is None:
            return True    
        if root.val >= right or root.val <= left:
            return False 
        return self.find(root.left, left, root.val) and self.find(root.right, root.val, right)
    
    def isValidBST(self, root):  
        return self.find(root, float("-inf"), float("inf"))
        